### PR TITLE
✨ Add Support for Optical Sensor

### PR DESCRIPTION
**Pull request by WillXuCodes**
### Summary:
Setup the C/Cpp API for the OpticalSensor.

### Motivation:
With the rotation sensor being released, PROS should support it since users already have their hands on it.

### Test Plan:

- [x] Test C/Cpp API with optical sensor

### CODE REVIEW

**Comment by kunwarsahni01** [comment]
Tested with hardware and all functions appear to be working properly.

File: include/pros/optical.h [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+namespace c {
+#endif
+
+typedef enum optical_direction_e {
+	NO_GESTURE = 0,
+	UP = 1,
+	DOWN = 2,
+	RIGHT = 3,
+	LEFT = 4,
+} optical_direction_e_t;
```
Comment (nathan-moore) [inline include/pros/optical.h:41]: consider making these lower case. All upper case is usually reserved for macros.

File: include/pros/optical.hpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+#define _PROS_OPTICAL_HPP_
+
+#include <stdbool.h>
+#include <cstdint>
+#include "pros/optical.h"
+
+namespace pros {
+class Optical {
+	public:
+  /**
```
Comment (nathan-moore) [inline include/pros/optical.hpp:29]: nit: looks like you need a formatting pass.

File: include/pros/optical.hpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	std::int32_t disable_gesture();
+
+  /**
+	 * Gets the port number of the Optical Sensor.
+	 *
+	 * \return The Optical Sensor's port number.
+	 */
+	std::uint8_t get_port();
+
+  protected:
```
Comment (nathan-moore) [inline include/pros/optical.hpp:226]: We usually make this private, though I agree it really should be protected. We should probably go through and do a consistency check.

File: include/pros/optical.hpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	 *
+	 * This function uses the following values of errno when an error state is
+	 * reached:
+	 * ENXIO - The given value is not within the range of V5 ports (1-21).
+	 * ENODEV - The port cannot be configured as an Optical Sensor
+	 *
+	 * \return hue value if the operation was successful or PROS_ERR_F if the operation
+	 * failed, setting errno.
+	 */
+	double get_hue();
```
Comment (nathan-moore) [inline include/pros/optical.hpp:57]: This is a question for @HotelCalifornia. We usually make these virtual, but I'm not entirely sure why.

File: src/devices/vdml_optical.c [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	vexDeviceOpticalRgbGet(device->device_info, &rgb);
+	rtn.red = rgb.red;
+	rtn.green = rgb.green;
+	rtn.blue = rgb.blue;
+	rtn.brightness = rgb.brightness;
+	return_port(port - 1, rtn);
+}
+
+#define RAW_ERR_INIT \
+	{ .clear = OPT_GESTURE_ERR, .red = OPT_GESTURE_ERR, .green = OPT_GESTURE_ERR, .blue = OPT_GESTURE_ERR }
```
Comment (nathan-moore) [inline src/devices/vdml_optical.c:76]: Is there a reason why these aren't pros error?

File: version [hunk 6]
```diff
@@ -1 +1 @@
-3.3.0
+3.3.0
```
Comment (HotelCalifornia) [inline version:1]: add the newline back so this isn't tracked as a change

File: src/devices/vdml_optical.c (hunk 5 above)
Comment (WillXuCodes) [reply src/devices/vdml_optical.c:76]: They aren't due to PROS_ERR being int32, and these values are int8 and int16

Comment (nathan-moore) [reply src/devices/vdml_optical.c:76]: They're uin32_t on the pros side?

Comment (HotelCalifornia) [reply src/devices/vdml_optical.c:76]: ah yeah they are for this struct, but not the other one

File: include/pros/optical.hpp (hunk 3 above)
Comment (kunwarsahni01) [reply include/pros/optical.hpp]: So should this be private for this PR?

File: include/pros/optical.hpp (hunk 4 above)
Comment (kunwarsahni01) [reply include/pros/optical.hpp:58]: Should this be virtual for this PR?

Comment (nathan-moore) [reply include/pros/optical.hpp:58]: Private and non-virtual are the defensive choices. We can do that and see if anyone asks for the need to extend them.

Comment (nathan-moore) [reply include/pros/optical.hpp:58]: private and all virtual is what we have done, so just do that for now.

File: include/pros/optical.h [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+ * 4 = left
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of V5 ports (1-21).
+ * ENODEV - The port cannot be configured as an Optical Sensor
+ *
+ * \param port
+ *        The V5 Optical Sensor port number from 1-21
+ * \return gesture value if the operation was successful or PROS_ERR if
```
Comment (nathan-moore) [inline include/pros/optical.h:213]: > or PROS_ERR if

This looks like ub: https://stackoverflow.com/questions/366017/what-is-the-size-of-an-enum-in-c. Looks like you need to make PROS_ERR part of the enum.

Comment (nathan-moore) [reply include/pros/optical.h:213]: Practically though, if it works now it's probably fine for now, though it's a simple fix.

File: include/pros/optical.h [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+ * Enable gesture detection on the sensor
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of V5 ports (1-21).
+ * ENODEV - The port cannot be configured as an Optical Sensor
+ *
+ * \param port
+ *        The V5 Optical Sensor port number from 1-21
+ */
```
Comment (nathan-moore) [inline include/pros/optical.h:243]: return comment?

File: include/pros/optical.h [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+	uint32_t blue;
+} optical_raw_s_t;
+
+typedef struct optical_gesture_s {
+	uint8_t udata;
+	uint8_t ddata;
+	uint8_t ldata;
+	uint8_t rdata;
+	uint8_t type;
+	uint8_t pad;
```
Comment (nathan-moore) [inline include/pros/optical.h:57]: Do you know if this is actual data or making the implicit padding explicit? If you don't, I wouldn't worry about it, as it's used for the low level "raw" apis. Also, if we have it, some docs on the individual fields would be nice.

**Review by nathan-moore — APPROVED** [review]
Overall LGTM

File: include/pros/optical.h [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+	uint16_t count;
+	uint32_t time;
+} optical_gesture_s_t;
+
+/**
+ * Get the detected color hue
+ *
+ * This is not avaliable if gestures are being detected. Hue has a
```
[the remark this answers is not in the document]
Comment (kunwarsahni01) [reply include/pros/optical.h:67]: It returns 0 when in gesture mode.

File: include/pros/optical.h (hunk 9 above)
Comment (kunwarsahni01) [reply include/pros/optical.h:59]: We don't have docs on the individual fields at this time.

**Review by nathan-moore — APPROVED** [review]

File: include/pros/optical.h [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+/**
+ * Get the most recent gesture data from the sensor
+ *
+ * Gestures will be cleared after 500mS
+ * 0 = no gesture
+ * 1 = up (towards cable)
+ * 2 = down
+ * 3 = right
+ * 4 = left
```
Comment (HotelCalifornia) [inline include/pros/optical.h:208]: this should be documented by the enum itself, not here

**Review by HotelCalifornia — COMMENTED** [review]
as always, my comments about the C documentation apply to the C++ header also

File: include/pros/optical.hpp [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	 * ENODEV - The port cannot be configured as an Optical Sensor
+	 *
+	 * \return The Error code encountered
+	 */
+	virtual std::int32_t disable_gesture();
+
+	/**
+	 * Gets the port number of the Optical Sensor.
+	 *
+	 * \return The Optical Sensor's port number.
```
[the remark this answers is not in the document]
Comment (kunwarsahni01) [reply include/pros/optical.hpp:223]: What are you referring to here? There is no `get_port` in the C file...

[the remark this answers is not in the document]
Comment (HotelCalifornia) [reply include/pros/optical.hpp:225]: 🤔 i swear there were two comments that both said "the error encountered" or something like that

**Review by HotelCalifornia — DISMISSED** [review]
lgtm

**Review by HotelCalifornia — APPROVED** [review]

**Pull request merged**